py/pysparkling/context.py

```diff
@@ -92,41 +92,58 @@ def __do_init(self, spark_session):
         self.is_initialized = False
 
     @staticmethod
-    def getOrCreate(spark, conf=None, **kwargs):
-        """
-         Get existing or create new H2OContext based on provided H2O configuration. If the conf parameter is set then
-         configuration from it is used. Otherwise the configuration properties passed to Sparkling Water are used.
-         If the values are not found the default values are used in most of the cases. The default cluster mode
-         is internal, ie. spark.ext.h2o.external.cluster.mode=false
+    def __default_h2o_connect(h2o_context, **kwargs):
+        return h2o.connect(ip=h2o_context._client_ip, port=h2o_context._client_port, **kwargs)
+
 
-         param - Spark Context or Spark Session
-         returns H2O Context
+    @staticmethod
+    def getOrCreate(spark, conf=None, pre_create_hook=None, h2o_connect_hook=__default_h2o_connect, **kwargs):
+        """
+        Get existing or create new H2OContext based on provided H2O configuration. If the conf parameter is set then
+        configuration from it is used. Otherwise the configuration properties passed to Sparkling Water are used.
+        If the values are not found the default values are used in most of the cases. The default cluster mode
+        is internal, ie. spark.ext.h2o.external.cluster.mode=false
+
+        :param spark: Spark Context or Spark Session
+        :param conf: H2O configuration as instance of H2OConf
+        :param pre_create_hook:  hook to reconfigure conf on given spark context
+        :param h2o_connect_hook:  hook which realize connection of h2o client
+        :param kwargs:  additional parameters which are passed to h2o_connect_hook
+        :return:  instance of H2OContext
         """
 
+        # Get spark session
         spark_session = spark
         if isinstance(spark, SparkContext):
             warnings.warn("Method H2OContext.getOrCreate with argument of type SparkContext is deprecated and " +
                           "parameter of type SparkSession is preferred.")
             spark_session = SparkSession.builder.getOrCreate()
+        # Get H2OConf
+        if conf is not None:
+            selected_conf = conf
+        else:
+            selected_conf = H2OConf(spark_session)
+
+        # Call pre_create hook
+        if pre_create_hook:
+            pre_create_hook(spark_session, selected_conf)
```

Comment:
We are passing H2OConf to steam/different service here as an argument - that probably means that steam will need pysparkling dependency

Reply:
yes

Reply:
https://github.com/h2oai/h2o-secure/commit/fde1bb711328a12d85de1c2e6948bcf406d27f50

Reply:
cool!